tools/src/main/java/org/apache/kafka/trogdor/agent/WorkerManager.java

```diff
@@ -0,0 +1,567 @@
+                        log.info("{}: Worker {} is halting.", nodeName, id);
+                    } else {
+                        log.info("{}: Worker {} is halting with error {}", nodeName, id, errorString);
+                    }
+                    stateChangeExecutor.submit(
+                        new HandleWorkerHalting(worker, errorString, false));
+                    return null;
+                }
+            });
+            try {
+                worker.taskWorker.start(platform, worker.status, haltFuture);
+            } catch (Exception e) {
+                stateChangeExecutor.submit(new HandleWorkerHalting(worker,
+                    "worker.start() exception: " + e.getMessage(), true));
+            }
+            stateChangeExecutor.submit(new FinishCreatingWorker(worker));
+        }
+    }
+
+    /**
+     * Handles a request to create a new worker.  Processed by the state change thread.
+     */
+    class CreateWorker implements Callable<Worker> {
+        private final String id;
+        private final TaskSpec spec;
+        private final long now;
+
+        CreateWorker(String id, TaskSpec spec, long now) {
+            this.id = id;
+            this.spec = spec;
+            this.now = now;
+        }
+
+        @Override
+        public Worker call() throws Exception {
+            Worker worker = workers.get(id);
+            if (worker != null) {
+                log.info("{}: Task ID {} is already in use.", nodeName, id);
+                return null;
+            }
+            worker = new Worker(id, spec, now);
+            workers.put(id, worker);
+            log.info("{}: Created a new worker for task {} with spec {}", nodeName, id, spec);
+            return worker;
+        }
+    }
+
+    /**
+     * Finish creating a Worker.  Processed by the state change thread.
+     */
+    class FinishCreatingWorker implements Callable<Void> {
+        private final Worker worker;
+
+        FinishCreatingWorker(Worker worker) {
+            this.worker = worker;
+        }
+
+        @Override
+        public Void call() throws Exception {
+            switch (worker.state) {
+                case CANCELLING:
+                    log.info("{}: Worker {} was cancelled while it was starting up.  " +
+                        "Transitioning to STOPPING.", nodeName, worker.id);
+                    worker.transitionToStopping();
+                    break;
+                case STARTING:
+                    log.info("{}: Worker {} is now RUNNING.  Scheduled to stop in {} ms.",
+                        nodeName, worker.id, worker.spec.durationMs());
+                    worker.transitionToRunning();
+                    break;
+                default:
+                    break;
+            }
+            return null;
+        }
+    }
+
+    /**
+     * Handles a worker halting.  Processed by the state change thread.
+     */
+    class HandleWorkerHalting implements Callable<Void> {
```

Comment:
Hmm.  I think it's accurate to call this a "handler" since it is not initiating a halt.  It is handling a worker that halted itself for whatever reason.  In contrast, StopWorker actually does stop the worker (it's the backend for `/agent/workers/stop`)